driver-core/src/main/java/com/datastax/driver/core/TypeCodec.java

```diff
@@ -29,357 +29,301 @@
-                    valueMap.put(valueEntry.getKey(), new MapCodec(codec, valueEntry.getValue(), protocolVersion));
-                primitiveMapsCodecs.put(type, valueMap);
-            }
-        }
-
-        private static final PrimitiveCollectionCodecs primitiveCollectionCodecsV2 = new PrimitiveCollectionCodecs(ProtocolVersion.V2);
-        private static final PrimitiveCollectionCodecs primitiveCollectionCodecsV3 = new PrimitiveCollectionCodecs(ProtocolVersion.V3);
-        private static final PrimitiveCollectionCodecs primitiveCollectionCodecsV4 = new PrimitiveCollectionCodecs(ProtocolVersion.V4);
-
-        static PrimitiveCollectionCodecs forVersion(ProtocolVersion version) {
-            // This happens during protocol negociation, when the version is not known yet.
-            // Use the smallest supported version, which is enough for what we need to do at this stage.
-            if (version == null)
-                version = ProtocolVersion.V1;
-
-            switch (version) {
-                case V1:
-                case V2:
-                    return primitiveCollectionCodecsV2;
-                case V3:
-                    return primitiveCollectionCodecsV3;
-                case V4:
-                    return primitiveCollectionCodecsV4;
-                default:
-                    throw version.unsupported();
-            }
-        }
+    protected TypeCodec(DataType cqlType, TypeToken<T> javaType) {
+        this.cqlType = cqlType;
+        this.javaType = javaType;
     }
 
+    /**
+     * Return the Java type that this codec deserializes to and serializes from.
+     * @return The Java type this codec deserializes to and serializes from.
+     */
+    public TypeToken<T> getJavaType() {
+        return javaType;
+    }
 
-    private TypeCodec() {}
-
-    public abstract T parse(String value);
-    public abstract String format(T value);
+    /**
+     * Return the CQL type that this codec deserializes from and serializes to.
+     * @return The Java type this codec deserializes from and serializes to.
+     */
+    public DataType getCqlType() {
+        return cqlType;
+    }
 
-    public abstract ByteBuffer serialize(T value);
-    public abstract T deserialize(ByteBuffer bytes);
+    /**
+     * Serialize the given value according to the CQL type
+     * handled by this codec.
+     * @param value An instance of T
+     * @param protocolVersion the protocol version to use when serializing
+     * {@code bytes}. In most cases, the proper value to provide for this argument
+     * is the value returned by {@link ProtocolOptions#getProtocolVersion} (which
+     * is the protocol version in use by the driver).
+     * @return A {@link ByteBuffer} instance containing the serialized form of T
+     * @throws InvalidTypeException if the given value does not have the expected type
+     * @throws NullPointerException if {@code value} is {@code null}, of for collections,
```

Comment:
I updated `TypeCodec` so that `serialize` and `deserialize` now accept null values; I also got rid of nullity tests wherever applicable to let the codecs decide what to do with a null value.
However:
* The `accept(Object)` method still doesn't accept null values, because otherwise any codec could be selected;
* `format` and `parse` still don't allow nulls (I currently don't see any advantage in letting them do so);
* To keep compatibility with the legacy code, collection codecs never return null (they return an empty collection instead).